### PR TITLE
Make tree splitters public

### PR DESCRIPTION
**Pull request by smastelini**
As discussed in #476, currently the attribute observers, a.k.a. splitters, are picked by passing strings to the trees. Their parameters are set with dictionaries. This choice is not optimal and makes it difficult to document and add new splitters in the long run.

This PR makes the splitter selection similar to what is done with linear models' optimizers.

In summary, this PR brings to the table:

- [x] Rename: Attribute observer (AO) -> splitter.
- [x] Make splitters public and standardize their naming scheme.
- [x] Add a property to differentiate classification and regression splitters (suggestions for a better property name are welcomed).
- [x] Splitters can be passed as a parameter to the trees. They are "deepcopied" to new leaves as they are created.
- [x] Remove redundant portions of the tree code.
- [x] Change indirectly impacted methods (ARF variants).
- [x] Apply misc/minor improvements to the `tree` module, e.g., fix `mypy` issues and standardize some variable naming patterns. 
- [x] Improve documentation in `tree.splitter` and possible mentions to splitters in the trees' docstrings.

What still needs to be done:
- [x] ~Create a page in the documentation explaining and comparing different splits in standalone trees and ensembles thereof. `progressive_val_score` is a good option, of course, but we could also use Trackers (#471).~
- [x] Decide whether or not other internal parts of the `tree` module should be refactored/streamlined.
- [x] Ensure the final documentation renders nicely. :D

### CODE REVIEW

**Comment by smastelini** [comment]
Those changes bring joy to my heart :smile:

However, this PR is just a stub yet. Suggestions are welcomed.

@MaxHalford and @jacobmontiel, thanks for the suggestions in #476. Is that what you had in mind?

File: river/tree/_nodes/efdtc_nodes.py
Comment (smastelini) [inline river/tree/_nodes/efdtc_nodes.py:79]: Forgot to change to `splitter`. I will do that later.

File: river/tree/_nodes/efdtc_nodes.py
Comment (smastelini) [inline river/tree/_nodes/efdtc_nodes.py:123]: What is the correct choice for docstrings? `kwargs` or  `**kwargs`?

File: river/tree/_nodes/htr_nodes.py
Comment (smastelini) [inline river/tree/_nodes/htr_nodes.py:55]: Need to change this too.

File: river/tree/base_hoeffding_tree.py
Comment (smastelini) [inline river/tree/base_hoeffding_tree.py:97]: Did that to avoid `mypy` errors.

File: river/tree/splitter/__init__.py
Comment (smastelini) [inline river/tree/splitter/__init__.py:3]: This also could be named `BSTSplitter`. Some papers do that if I recall it correctly. The regression version "extends" this algorithm, so it is called "Extended Binary Search Tree {Splitter} Attribute Observer" -> `EBSTSplitter`.

Comments, @jacobmontiel?

File: river/tree/splitter/base_splitter.py
Comment (smastelini) [inline river/tree/splitter/base_splitter.py:12]: Review that

File: river/tree/splitter/qo_splitter.py
Comment (smastelini) [inline river/tree/splitter/qo_splitter.py:32]: The plan is to have something like that for each splitter. Adding documentation to this one was easy because I am the author hahaha

For the other ones, any help is welcomed! @jacobmontiel, you probably know more than I do about many of the other splitters. If you want and have time for that, I would be happy to collaborate on this documentation workload.

**Comment by MaxHalford** [comment]
Looking good!

**Review by smastelini — COMMENTED** [review]
I left some comments regarding my own changes. Some of them are reminders to myself, others are questions for further discussion. As I see it, the splitters will need a boost in documentation to facilitate their usage by the users.

File: river/tree/_nodes/efdtc_nodes.py
Comment (MaxHalford) [reply river/tree/_nodes/efdtc_nodes.py:123]: `kwargs`

File: river/tree/_nodes/efdtc_nodes.py
Comment (smastelini) [reply river/tree/_nodes/efdtc_nodes.py]: Changed in commit [d6931d9](https://github.com/online-ml/river/pull/488/commits/d6931d9e1632978aeb7276b9a13cff1ac2a3ec09).

File: river/tree/_nodes/efdtc_nodes.py
Comment (smastelini) [reply river/tree/_nodes/efdtc_nodes.py:123]: Thanks! I corrected the docstrings in commit [d6931d9](https://github.com/online-ml/river/pull/488/commits/d6931d9e1632978aeb7276b9a13cff1ac2a3ec09).

File: river/tree/_nodes/htr_nodes.py
Comment (smastelini) [reply river/tree/_nodes/htr_nodes.py]: Changed in commit [d6931d9](https://github.com/online-ml/river/pull/488/commits/d6931d9e1632978aeb7276b9a13cff1ac2a3ec09).

File: river/tree/splitter/base_splitter.py
Comment (smastelini) [reply river/tree/splitter/base_splitter.py]: Changed in commit [d6931d9](https://github.com/online-ml/river/pull/488/commits/d6931d9e1632978aeb7276b9a13cff1ac2a3ec09).

File: river/tree/splitter/qo_splitter.py
Comment (smastelini) [reply river/tree/splitter/qo_splitter.py:32]: I improved the documentation a bit in a recent commit.

**Comment by smastelini** [comment]
Since #471 is not ready yet, I think we can merge this PR when we are happy with the changes and wait for the `Tracks` to update our documentation page.

Again, I intend to provide some benchmarks of the splitters using synthetic datasets.

What do you think, @jacobmontiel and @MaxHalford?

**Comment by smastelini** [comment]
Hey @MaxHalford and @jacobmontiel, I added a page in the user guides that is a walkthrough in the tree module. I talked about the different tree models available, model inspection, memory management, and splitters. Hope that can help the users.

Your feedback is welcomed.

(Later today I'll add some basic coverage tests for the splitters)

**Comment by smastelini** [comment]
The failing tests are related to classifier chains. I believe they were recently updated. Pinging @MaxHalford.

**Comment by smastelini** [comment]
All set. If you require more changes before merging, please let me know, @jacobmontiel and @MaxHalford.

File: river/tree/__init__.py
Comment (MaxHalford) [inline river/tree/__init__.py:13]: Not sure we want to expose this base class, right?

File: river/tree/_nodes/base.py
Comment (MaxHalford) [inline river/tree/_nodes/base.py:11]: I would call this simply `SplitSuggestion` instead of `AttributeSplitSuggestion`, but your call!

**Review by MaxHalford — COMMENTED** [review]
Good job, I mean it! It looks like you put in a lot of work. The code looks much nicer IMO. And the fact that the parameters are not duplicated everywhere in the docs feels better too.

File: river/tree/__init__.py
Comment (smastelini) [reply river/tree/__init__.py:13]: It's kind of tricky. It has some important documentation, that's the only reason it's public right now. I'm not sure this is the best approach, but it was a way to avoid repetition.

File: river/tree/_nodes/base.py
Comment (smastelini) [reply river/tree/_nodes/base.py:11]: I like that. I am so used to this code that I didn't realize it was a long name till you mentioned it. I'll rename it tomorrow :)

File: river/tree/__init__.py
Comment (MaxHalford) [reply river/tree/__init__.py:13]: I'm not 100% but I think we can play with the `__doc__` attribute of classes. So you could have some string that contains the doc, and just add it the `__doc__` of the class you want to add the documentation to. I've never done it but I've seen it done in (serious) projects :)

File: river/tree/_nodes/base.py
Comment (smastelini) [reply river/tree/_nodes/base.py]: Renamed in [2e63165](https://github.com/online-ml/river/pull/488/commits/2e63165b10a2c12af7ada79bc606cfb5c44805ad).

File: river/tree/base_hoeffding_tree.py
Comment (jacobmontiel) [inline river/tree/base_hoeffding_tree.py:26]: I would rather keep the name as `BaseHoeffdinTree` as we do in other places (such as for ensembles). As previously mentioned, this class should not be public.

File: river/tree/extremely_fast_decision_tree.py
Comment (jacobmontiel) [inline river/tree/extremely_fast_decision_tree.py:55]: I would add: "This is an advanced option. Special care must be taken when choosing a different splitter." The same applies to other descriptions similar to this one.

File: river/tree/splitter/ebst_splitter.py
Comment (jacobmontiel) [inline river/tree/splitter/ebst_splitter.py:248]: In excess of caution, should we name this as `EBSTNode`?

File: river/tree/splitter/base_splitter.py
Comment (jacobmontiel) [inline river/tree/splitter/base_splitter.py:48]: I suggest keeping it as `target_val`, for consistency with the `update` method.

File: river/tree/splitter/exhaustive_splitter.py
Comment (jacobmontiel) [inline river/tree/splitter/exhaustive_splitter.py:175]: Same as before, what about `ExhaustiveNode`? 

This is optional.

File: river/tree/splitter/histogram_splitter.py
Comment (jacobmontiel) [inline river/tree/splitter/histogram_splitter.py:86]: Could you elaborate on this? Code after `if` is no longer part of the loop?

File: river/tree/splitter/nominal_class_splitter.py
Comment (jacobmontiel) [inline river/tree/splitter/nominal_class_splitter.py:9]: The new names are kind of confusing. As they got shorter what about: `NominalSplitterClassif` and `NominalSplitterReg` respectively?

**Review by jacobmontiel — CHANGES_REQUESTED** [review]
Awesome work @smastelini 

I really like how the new naming and structure tidies-up the code. Nice catch on those unnecessary parts of the code. The QO splitter refactoring is also nice. Not sure if it is listed in the PR description.

I left a couple of comments and changes. In general, the PR looks good.

Let me know if you need further information.

File: river/tree/splitter/__init__.py
Comment (smastelini) [reply river/tree/splitter/__init__.py:3]: Let's keep the name ExhaustiveSplitter since this is the common choice in the literature.

File: river/tree/__init__.py
Comment (smastelini) [reply river/tree/__init__.py:13]: That's a good point. Keep in mind, however, that we do have some other abstract classes that are exposed in the documentation. For example, `optim.Optimizer`. I am not sure whether or not this was intended.

> On the other hand, it is unlikely that people will find or search for the documentation in a different class.

That is a valid point. Currently, we point to `tree.HoeffdingTree` from the other trees. Something like: "for more information check `tree.HoeffdingTree`".

Again, I would rather keep the plethora of somewhat obscure (yet so useful) parameters documented in the base class than exposing them everywhere and confusing the user with too many choices to pick from. A page in the documentation could do the work, but I find it strange to document parameters outside of a docstring. If we opt for the first option and replicate the parameter information everywhere, we can follow Max's suggestion.

[`matplotlib`](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.plot.html) also relies a lot on `**kwargs`. Maybe we could provide some pieces of information about what the user can do and then point to the base class. Hence, we balance the amount of the "data overhead" for the user.

File: river/tree/_nodes/base.py
Comment (smastelini) [reply river/tree/_nodes/base.py:83]: I see your point. However, as we don't make copies of the internal structures when returning the properties, one could still do something like:

```python
n_stats = node.stats  # property access

n_stats['class_name'] = 100000

print(node.stats['class_name'])
>> 100000
```
and the node's statistics would still be changed, right? :P

But nodes are a private module that is only accessed by the trees internally. So, it is (probably) fine to keep the nodes' attributes public. What do you think?

File: river/tree/base_hoeffding_tree.py
Comment (smastelini) [reply river/tree/base_hoeffding_tree.py:26]: Regardless of making this class public or not, I still do prefer shorter names hahaha

I think we need to define which pattern we are going to use across the codebase. For instance, we have:

- `base.Classifier` (and its friends)
- `metrics.Metric`

but we also have:

- `ensemble.BaseForest`
- `base.WrapperMixin`

I don't think this is a pressing matter whatsoever, but it would be nice to use the same pattern everywhere. Personally, I prefer using shorter names as much as possible: Classifier, Metric, Forest, HoeffdingTree, etc.

What do you think, @jacobmontiel and @MaxHalford ?

File: river/tree/splitter/base_splitter.py
Comment (smastelini) [reply river/tree/splitter/base_splitter.py:48]: I decided to use `class_val` to make it clear that conditional probabilities only apply to the classification case.
What do you think? I am totally cool with both of them.

File: river/tree/splitter/nominal_class_splitter.py
Comment (smastelini) [reply river/tree/splitter/nominal_class_splitter.py]: Yes! Much better haha

What was I thinking? Nominal Reg/Class Splitter, LOL

Changed in commit [b71f853](https://github.com/online-ml/river/pull/488/commits/b71f853f5379e0e61b1b70902babc2ff3a4f3ce0).

File: river/tree/_attribute_test/__init__.py
Comment (smastelini) [reply river/tree/_attribute_test/__init__.py]: Wonderful! Renamed in commit [9aeda54](https://github.com/online-ml/river/pull/488/commits/9aeda5439cde9080d45254a45d4044aa885e36d8).

File: river/tree/splitter/ebst_splitter.py
Comment (smastelini) [reply river/tree/splitter/ebst_splitter.py]: I agree with you, it is better to make things clear. Renamed in commit [3b4fed1](https://github.com/online-ml/river/pull/488/commits/3b4fed13fee51b7f77a3b5226ac24f27894f44ff).

File: river/tree/splitter/exhaustive_splitter.py
Comment (smastelini) [reply river/tree/splitter/exhaustive_splitter.py]: Renamed in commit [3b4fed1](https://github.com/online-ml/river/pull/488/commits/3b4fed13fee51b7f77a3b5226ac24f27894f44ff).

File: river/tree/_nodes/arf_htc_nodes.py
Comment (smastelini) [reply river/tree/_nodes/arf_htc_nodes.py:120]: My bad. I'll fix it!

File: river/tree/_nodes/arf_htc_nodes.py
Comment (smastelini) [reply river/tree/_nodes/arf_htc_nodes.py]: I don't even know anymore :D. It doesn't make much sense, so I removed this portion (commit [6430a0e](https://github.com/online-ml/river/pull/488/commits/6430a0e4df6ce0fc2f07fc228384a93716c2f1da)). Thanks for mentioning that.

File: river/tree/_nodes/hatc_nodes.py
Comment (smastelini) [reply river/tree/_nodes/hatc_nodes.py]: Nice! Added in [a0ff412](https://github.com/online-ml/river/pull/488/commits/a0ff4122f2263214bf3cff25b9412655b0caa3e7)

File: river/tree/splitter/histogram_splitter.py
Comment (smastelini) [reply river/tree/splitter/histogram_splitter.py:86]: This was a bug :/

We need to iterate over all classes in `pre_split_dist` before evaluating the resulting post-split distribution. We didn't have tests for the different splitters until now so that one escaped my attention.

File: river/tree/_nodes/base.py
Comment (jacobmontiel) [reply river/tree/_nodes/base.py:83]: Agree. Nodes are internal classes and it is nice to avoid the overhead from the properties. On the other hand, we must be careful to protect the nodes when exposing results to the user at the estimator level.

File: river/tree/extremely_fast_decision_tree.py
Comment (smastelini) [reply river/tree/extremely_fast_decision_tree.py:56]: I added this comment in the last commit!

File: river/tree/splitter/base_splitter.py
Comment (smastelini) [reply river/tree/splitter/base_splitter.py]: I updated all the occurrences to `target_val` for consistency.

File: river/tree/__init__.py
Comment (smastelini) [reply river/tree/__init__.py:13]: @MaxHalford, would it be possible to add this portion of the documentation at the "index" page of the tree module? We do not have that now, but it could be a viable solution.

I am also convinced that `HoeffdingTree` should not be public.

Comment (MaxHalford) [reply river/tree/__init__.py:13]: Yes we can and should documentation in the `__init__.py` file if that's what you mean.

File: river/tree/base_hoeffding_tree.py
Comment (MaxHalford) [reply river/tree/base_hoeffding_tree.py:26]: Actually my opinion is that we should not add the `Base` prefix. In any case these classes are not to be user-facing.

File: river/tree/__init__.py
Comment (smastelini) [reply river/tree/__init__.py:13]: Sounds like a plan. Probably I will not use the same notation, (the `Parameters` section, for instance) but it will do the work!

**Comment by smastelini** [comment]
I updated the documentation in the last split to make `tree.HoeffdingTree` hidden again. @jacobmontiel, please let me know if the documentation placement is satisfactory now :D

**Comment by smastelini** [comment]
Disclaimer: the failing test is unrelated to this PR.

**Review by jacobmontiel — APPROVED** [review]
Thanks for the last changes @smastelini 
Issues found will be addressed in another PR.

**Comment by smastelini** [comment]
Thanks for all your help and support, @jacobmontiel and @MaxHalford!

**Pull request merged**
